Scalefactor development

JulesVandenbroeck opened the pull request:
### overview

This pull request encoperates a lot of changes to how scale factors and uncertainties are derived in columnflow. To see a guide on most uncertainties/scale factors see the scalefactors building block guide [here](https://github.com/GhentAnalysis/columnflow/blob/scalefactor-development/docs/user_guide/building_blocks/scalefactors.md). An overview of the changes are:
- **NEW** Implementation of fixed working point b-tagging scale factors as a weight producer together with custom tasks `cf.BTagEfficiency` and `cf.BTagEfficiencyPlot` to calculate and plot b-tagging efficiencies.
-  **NEW** Implementation of trigger scale factors as a weight producer together with custom tasks `cf.TriggerScaleFactors`, `PlotTriggerScaleFactors1D`, `PlotTriggerScaleFactors2D`, and `PlotTriggerEfficiencies1D`  to calculate and plot the trigger scale factors and trigger efficiencies.
-  **NEW**  parton shower uncertainties as a weight producer
-  **UPDATE**  re-organization of the lepton scale factor implementation with custom `LeptonSFConfig` to configure the lepton scale factor weight producer.
- **NEW** documentation for scale factors and uncertainties in columnflow
- **UPDATE** `pdf_weights` producer now also produces $\alpha_s$ variation weights
- **UPDATE** `pu_weight` producer now has the option to apply a weight threshold which sets the `pu_weights` and `mc_weight` columns to 0 if any weight variation is above the threshold. If this option is used the `pu_weight` producer should also be run in Selection to remove any event with weights above the threshold from the weight normalization.
- **UPDATE** The GhentAnalysis analysis template is updated to include the updated use of lepton scale factors and includes both b-tagging scale factors and parton shower uncertainties (trigger scale factors not yet added)

__________________________

### User changes
**Required**
- None

**Optional**
- move to updated lepton scale factor implementation. For an example see the template. (relevant files are [lepton_settings.py](https://github.com/GhentAnalysis/columnflow/blob/scalefactor-development/analysis_templates/ghent_template/__cf_module_name__/config/settings/lepton_settings.py), [shifts.py](https://github.com/GhentAnalysis/columnflow/blob/180ff08c86f790db4bb71fbf7a661fa9fd10a4c9/analysis_templates/ghent_template/__cf_module_name__/config/shifts.py#L40-L64) and [production/weights.py](https://github.com/GhentAnalysis/columnflow/blob/180ff08c86f790db4bb71fbf7a661fa9fd10a4c9/analysis_templates/ghent_template/__cf_module_name__/production/weights.py#L130))
- add parton shower uncertainties to the weight producer of your analysis. For an example see the template. (relevant files are [production/weights.py](https://github.com/GhentAnalysis/columnflow/blob/180ff08c86f790db4bb71fbf7a661fa9fd10a4c9/analysis_templates/ghent_template/__cf_module_name__/production/weights.py#L98-L101), [shifts.py](https://github.com/GhentAnalysis/columnflow/blob/180ff08c86f790db4bb71fbf7a661fa9fd10a4c9/analysis_templates/ghent_template/__cf_module_name__/config/shifts.py#L73-L76) and [production/weights.py](https://github.com/GhentAnalysis/columnflow/blob/180ff08c86f790db4bb71fbf7a661fa9fd10a4c9/analysis_templates/ghent_template/__cf_module_name__/production/weights.py#L98-L101) also [here](https://github.com/GhentAnalysis/columnflow/blob/180ff08c86f790db4bb71fbf7a661fa9fd10a4c9/analysis_templates/ghent_template/__cf_module_name__/production/weights.py#L146-L167))
- add the b-tagging scale factor producer to your analysis. For an example see the template. (relevant files are [btagging_settings.py](https://github.com/GhentAnalysis/columnflow/blob/scalefactor-development/analysis_templates/ghent_template/__cf_module_name__/config/settings/btagging_settings.py), [selection/default.py](https://github.com/GhentAnalysis/columnflow/blob/180ff08c86f790db4bb71fbf7a661fa9fd10a4c9/analysis_templates/ghent_template/__cf_module_name__/selection/default.py#L238-L245) and [production/weights.py](https://github.com/GhentAnalysis/columnflow/blob/180ff08c86f790db4bb71fbf7a661fa9fd10a4c9/analysis_templates/ghent_template/__cf_module_name__/production/weights.py#L133-L135) )
__________________________

### Future development

- [x] pdf uncertainties are not following [TOP-recommendations](https://twiki.cern.ch/twiki/bin/viewauth/CMS/TopSystematics#PDF). The possibility to produce all pdf variations separately is available and can be used to follow TOP-recommendations.
- [x] $\alpha_s$ uncertainties are not yet included in columnflow
- [x] The ability to include the production of plots in the requirements of scale factor `producers`  like `trigger_scale_factors` and `fixed_wp_btag_weights` is currently missing.

Commit:
custom electron_weight function in columnflow/production/cmsGhent/electron.py

JulesVandenbroeck committed:
Btagging scale factors in scalefactor-development (#51)

* producers for btags, effiencies histograms, and sf. task for efficiency calculation

* remove btag_weights from law.cfg (not required to be resolved in task)

* apply also jet selection before histogramming

* use last cmd line process for labeling

* - calculation for multiple working point
- added correlated and uncorrelated

* fixed mistakes

fixed mistakes

---------

Co-authored-by: maadcoen <maarten.de.coen@cern.ch>

Commit:
add custom muon_weight producer & change weight initialization

Commit:
Update of JECR calibration to have flexibility in jet naming

maadcoen committed:
reorganize b-tagging

maadcoen committed:
fix bugs

fix bugs

Commit:
[tmp] Commit before bundling

maadcoen committed:
trigger sf initial commit

maadcoen committed:
split into plotting and calculation tasks

maadcoen committed:
fix bugs

maadcoen committed:
fix linting

maadcoen committed:
add koopman uncertainties for scale factors

maadcoen committed:
resolve default reference trigger

maadcoen committed:
fix linting

maadcoen committed:
add cmsGhent tasks to law

maadcoen committed:
HLT. only with trigger_vr

maadcoen committed:
fix issues

Commit:
various up and down variations disabled, needs to be customized

maadcoen committed:
fix issues

maadcoen committed:
fix linting

maadcoen committed:
refactor dependencies

Commit:
adaption of cms/jets.py that includes the possibility to use PuppiMET

Commit:
fix training comma

Commit:
fix trailing comma

maadcoen committed:
fix bugs

maadcoen committed:
store figures in same directory as scale factors

Commit:
changes to virtual environment requirements to add statsmodels module

Commit:
removed plot_2d config producers

JulesVandenbroeck commented:
Made changes to the sandboxes used for `SelectionEfficiencyHistMixin`. It now uses `bash::$CF_BASE/sandboxes/venv_selection_eff.sh` which has the module `statsmodels` installed. The task `BTagEfficiencyPlot` sandbox is changed again to `dev_sandbox(law.config.get("analysis", "default_columnar_sandbox"))` as leaving it out made the task run locally (needs to be fixed?).

Commit:
output key in btag producer did not match output key in btag efficiency task

Commit:
added sources feature to BTagSFConfig class

Commit:
changed column names to have _up/down last to match shifts. Also made systematics flexible using BTagSFConfig.sources

Commit:
included the use of .json files instead of .gs files only as correctionlib files

JulesVandenbroeck commented:
@maadcoen 
The `columnflow/production/cmsGhent/trigger_scale_factors.py` file is missing with in it the histograms that need to be producer during `cf.SelectEvents` and the producer to apply the trigger scale factors + uncertainties. I assume the code for this does exist and you have it on a local branch?

Commit:
beginning of trigger producer and histogramming; work in progress

Commit:
beginning of trigger producer and histogramming; work in progress

Commit:
beginning of trigger producer and histogramming; fixes

maadcoen committed:
allow to specify btagging efficiencies using external files rather than custom task

maadcoen committed:
make location btag eff and sf configurable

Commit:
addition of DatasetMixin Task and changes to trigger scalefactors

maadcoen committed:
deal with cases with negative counts: warning and set to zero

Commit:
linting fixes

maadcoen committed:
move trigger config back to producer, remove new method, add copy method

maadcoen committed:
fix warning in koopman test

maadcoen committed:
remove dangling import

Commit:
added 'get_btag_sf' attribute to def jet_btag as it is required for setup.

Commit:
linting fixes

Commit:
updated get_trigger_config and TriggerSFConfig class

Commit:
linting fixes

maadcoen committed:
add aux, require presence of get_trigger_config

JulesVandenbroeck commented on columnflow/tasks/cmsGhent/trigger_scale_factors.py:
```diff
@@ -0,0 +1,819 @@
+
+            # broad cast projected hist back to nominal binning
+            vr_idx = nom_centr_sf.axes.name.index(vr.name)
+            bias = np.expand_dims(corr_bias[vr.name], np.delete(np.arange(nom_centr_sf.ndim), vr_idx).tolist())
```
I don't quite understand what is going on here. Goal is to expand the correlation bias in 1D efficiencies (hist) to a 2d `np.Array`? This is not working as intended I believe.

maadcoen replied:
the nominal SF have a number of dimensions, corresponding to the number of binning variables. This can be, for instance, (p_T, eta, njet)

In this part of the code, I projected everything on one of these binning variables, to calculate the correlation bias as a function of this variable only. Could be that there is only a significant trend in the correlation observed as a function of p_T. 

This correlation bias is then added as an uncertainty to the nominal scale factors, assuming it then doesn't depend on the other binning variable. To do so, the shape of `corr_bias[vr.name]` must be broadcastable to `nom_centr_sf`. So if `vr.name == pt`, then we have to broadcast it to `(n_pt_bins, 1, 1)`, if `vr.name == eta`, then to `(1, n_eta_bins, 1)`, if `vr.name == njet`, then `(1, 1, n_njet_bins)`.

 `np.arange(nom_centr_sf.ndim)` gives you the indices of all axes, `np.delete(np.arange(nom_centr_sf.ndim), vr_idx)` removes the index of the variable at hand, `np.expand_dims(corr_bias[vr.name], np.delete(np.arange(nom_centr_sf.ndim), vr_idx).tolist())` then adds dimensions of lenght 1 at the remaining indices.

I don't see what could not be working as intended, but maybe I am missing something?

maadcoen committed:
missing get_btag_sf

maadcoen committed:
remove redundant parentheses

Commit:
systematic name and up/down variations in correctionlib file are now customizable using additional variables in the producer called 'syst_key' and 'postfixes'.

Commit:
fixes to btagging producer including addition of get_btag_sf attribute and removal of processes parameter being set in requirements.

Commit:
fix linting

maadcoen committed:
change default trigger config tags

maadcoen committed:
fix some bugs

maadcoen committed:
fix plotting tasks

maadcoen committed:
fix trig hist plot

maadcoen committed:
replaced non existing np.flatten

maadcoen committed:
by default al b-tagging working points

maadcoen committed:
lepton weight producers and config

maadcoen committed:
fix bugs

maadcoen committed:
fix linting

maadcoen committed:
spelling mistake

maadcoen committed:
fix case where dictionary is given to systematics

JulesVandenbroeck commented on columnflow/production/cmsGhent/lepton.py:
```diff
@@ -0,0 +1,329 @@
+    """
+
+    variable_map = self.input_func(events)
+    if self.mask_func or mask:
```
Would simplify this to a single line that only looks at the `mask_func` given in the config and no longer allow for additional mask being given during run-time unless I am missing a specific use case?
`mask = self.mask_func(events) if self.mask_func else Ellipsis`

maadcoen committed:
fix bugs in lepton.py

maadcoen committed:
fix bug in mask evaluation

maadcoen committed:
mask default ellipsis to none

maadcoen committed:
Update lepton.py: redefinition removed

maadcoen committed:
Update trigger_scale_factors.py:fix bug in stat error calculation

maadcoen committed:
remove redundant updates of the scale_factors["nominal"] dict

Commit:
update on examples of LeptonWeightConfig and a bundle_lepton_weights function that loops over all configs to derive weights

Commit:
linting fix 304:73

maadcoen committed:
move more settings to the config

maadcoen committed:
config naming configurable (in case of collision)

maadcoen committed:
add trigger bundle

maadcoen committed:
get_trigger_config > trigger_config, and missing mc_only

maadcoen committed:
add event mask function to the trigger config

maadcoen committed:
fix mistakes in uses

maadcoen committed:
fix mistake in default sf name

maadcoen committed:
Update trigger.py: get_trigger_config > trigger_config

maadcoen committed:
import error fixed

maadcoen committed:
Update trigger.py: fix init

maadcoen committed:
Update trigger.py: remove redundant self argument in lambda

maadcoen committed:
Update lepton.py: fix typo

maadcoen committed:
Update trigger.py: fix mistake in derivation

maadcoen committed:
Update trigger.py: sf_name now str, not function

maadcoen committed:
Update trigger.py: undefined event_mask_uses > defined uses

maadcoen committed:
refactor scale factor task:
extended config (__init__.py) used also in trigger task
moved original producers for sf and histograms to separate files
uncertainty "producers" (note: not like columnflow producers)

maadcoen committed:
simplify initialization of producers

maadcoen committed:
missing order import

maadcoen committed:
variable_insts > variables

maadcoen committed:
typo fixed

JulesVandenbroeck commented on columnflow/production/cmsGhent/trigger/sf_producer.py:
```diff
@@ -0,0 +1,150 @@
+            self.task,
+            datasets=self.datasets,
+            variables=self.variables,
+            trigger=self.tag,
```
@maadcoen  I believe this definition is outdated and the parameter should be `trigger_config` rather than `trigger` and `ref_trigger`

Commit:
changes to get_trigger_config that can now get a TriggerSFConfig. A TriggerSFConfig in config_inst.x.trigger_sf is now also prioritised over trigger_sfs

Commit:
bugfixes in requirements of trigger_scale_factors producer. Should be improved as duplicate information is given

Commit:
changed self.variables completion to use DotDict to add Trigger variables in case of double initialization and ordering.

Commit:
linting fixes

maadcoen committed:
simplify trigger tasks, remove variables arguments

maadcoen committed:
fix bugs

maadcoen committed:
plot prefix

maadcoen committed:
bug fix 2d plot

maadcoen committed:
fix plot hist

maadcoen committed:
fix bugs

maadcoen committed:
fix bugs in hist plots

maadcoen committed:
fix stat uncertainty

maadcoen committed:
fix bugs

maadcoen committed:
Update hist_producer.py

trigger bundle also on data

maadcoen committed:
bundle sf also on data

maadcoen committed:
add parton shower weights

maadcoen committed:
fill_hist moved

Commit:
linting fixes

Commit:
update of prepare_plot_config to prepare_stack_plot_config

Commit:
bug fix unrolled

Commit:
fix of MergeSelectionStats in selection_hists, removed exclude_params_forest_merge and tree_index

Commit:
fixes to plotting of Integer hist axes

Commit:
removed remaining breakpoints

Commit:
custom jet calibration is no longer needed because of upstream changes

JulesVandenbroeck committed:
Create scalefactors.md

Initial start of scale factor documentation

JulesVandenbroeck committed:
Update scalefactors.md

fixes of links

JulesVandenbroeck committed:
Update scalefactors.md

documentation for lepton scalefactors (missing documentation for shifts)

JulesVandenbroeck committed:
Update scalefactors.md

b-tagging scale factors documentation

Commit:
structural changes to the template to include scale factor development (no trigger scale factors yet)

Commit:
linting fix + documentation

Commit:
linting fixes in scale factor docs

Commit:
temporary change of create analysis columnflow branch

Commit:
small changes to README tutorial code

Commit:
template fixes, clean-up, and linting fixes to incorporate scale factor developments

Commit:
fixed_wp_btag_weights_init did not have a return in case of no dataset_inst

Commit:
linting fixes of docs

Commit:
linting fixes of docs

Commit:
undo of temporary change of create analysis columnflow branch

JulesVandenbroeck committed:
Merge branch 'GhentAnalysis/master' into scalefactor-development

maadcoen committed:
Update scalefactors.md

maadcoen committed:
Update scalefactors.md

maadcoen committed:
avoid calling producers in both produces and uses in bundle functions

Commit:
addition of alphas weights to pdf_weights

Commit:
fixes to bundle lepton weights

Commit:
linting fix

Commit:
fixes to template MuonWeightConfig

Commit:
update to pile-up weights that removes events with pu-weight above the threshold

maadcoen committed:
fix bug in trigger scale factor task

Commit:
changed pile-up weight to set all weights to 0

JulesVandenbroeck committed:
update of scalefactor development from GhentAnalysis/master (#85)

* fix of discrete_x variables plotting

* none protect jec branches

* Update jets.py

comment to 1 line of code

* mixin to avoid rerunning get_param_values needlessly (#83)

* mixin to avoid rerunning get_params needlessly

* timing code

* check cached params and add option to force repeated calls to get_param_values

---------

Co-authored-by: juvanden <jules.vandenbroeck@cern.ch>
Co-authored-by: Jan van der Linden <jan.vdlinden@cern.ch>
Co-authored-by: Maarten De Coen <52047931+maadcoen@users.noreply.github.com>

Commit:
changes from GhentAnalysis/fix_get_params_v2

Commit:
update plott_all function to allow for callable methods in plotting_config

Commit:
add parameter produce_plots in btag_weights producer to add BTagEfficiencyPlot task to its requirements

Commit:
fix to trigger scalefactor producer where TriggerScaleFactors task does no longer require variables

Commit:
remove ProcessesMixin from BTagEfficiencyPlot task and add VariablesMixin. Added dataset-group parameter for plotting label. Also create a dummy process to use for plotting with this label
